Error with too long output filenames on Windows

A user using Microsoft Windows 10 created this compilation:
```
[7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7] 
[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 30, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 30, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 60] 
3
```
which resulted in this output filename (379 characters for the file name alone, 405 for the full path):
```
7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_30b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_30b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_60b-x3.playlist.txt
```
which caused an error during the OS FS operation:
```
2023-08-30--13:09:46.649 - MainControl - Started compilation with: [7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7] [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 30, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 30, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 60] 3
2023-08-30--13:09:46.651 - CompilationGenerator - Progress: 0, Lese Daten ein
2023-08-30--13:09:46.652 - CompilationGenerator - Progress: 10, Lade Musikdateien 
2023-08-30--13:09:46.652 - CompilationGenerator - Progress: 20, Lade Pausenmusikdateien
2023-08-30--13:09:46.652 - CompilationGenerator - Progress: 30, Erzeuge Wiedergabeliste
2023-08-30--13:09:46.665 - unknown - Folgender Fehler trat auf.
C:\Users\USERNAME\Desktop\7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_30b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_30b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_60b-x3.playlist.txt (Die Syntax für den Dateinamen, Verzeichnisnamen oder die Datenträgerbezeichnung ist falsch)
2023-08-30--13:09:46.666 - IntervalMusicCompositorUI - ch.retorte.intervalmusiccompositor.compilation.CompilationException: Folgender Fehler trat auf.
C:\Users\USERNAME\Desktop\7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_30b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_30b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_60b-x3.playlist.txt (Die Syntax für den Dateinamen, Verzeichnisnamen oder die Datenträgerbezeichnung ist falsch)
                 at intervalmusiccompositor.core/ch.retorte.intervalmusiccompositor.compilation.CompilationGenerator.createPlaylistReport(CompilationGenerator.java:243)
                 at intervalmusiccompositor.core/ch.retorte.intervalmusiccompositor.compilation.CompilationGenerator.run(CompilationGenerator.java:96)
                 at java.base/java.lang.Thread.run(Thread.java:833) 
Caused by: java.io.FileNotFoundException: C:\Users\USERNAME\Desktop\7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_30b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_30b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_1b_7s_60b-x3.playlist.txt (Die Syntax für den Dateinamen, Verzeichnisnamen oder die Datenträgerbezeichnung ist falsch)
                 at java.base/java.io.FileOutputStream.open0(Native Method)
                 at java.base/java.io.FileOutputStream.open(FileOutputStream.java:293)
                 at java.base/java.io.FileOutputStream.<init>(FileOutputStream.java:235)
                 at java.base/java.io.FileOutputStream.<init>(FileOutputStream.java:184)
                 at intervalmusiccompositor.core/ch.retorte.intervalmusiccompositor.compilation.CompilationGenerator.createPlaylistReport(CompilationGenerator.java:237)
                 ... 2 more
```

According to https://learn.microsoft.com/en-us/windows/win32/fileio/naming-a-file?redirectedfrom=MSDN#maximum-path-length-limitation the maximum path length is 260:
> In editions of Windows before Windows 10 version 1607, the maximum length for a path is MAX_PATH, which is defined as 260 characters. In later versions of Windows, changing a registry key or using the Group Policy tool is required to remove the limit. See [Maximum Path Length Limitation](https://learn.microsoft.com/en-us/windows/win32/fileio/maximum-file-path-limitation) for full details.

The software should not exceed this limit in the Microsoft Windows version.

--- a/intervalmusiccompositor.app/src/main/resources/CHANGELOG.txt
+++ b/intervalmusiccompositor.app/src/main/resources/CHANGELOG.txt
@@ -4,6 +4,7 @@ retorte laboratories (retorte.ch) IntervalMusicCompositor change log
 202x-xx-xx  2.11.5 [not yet released]
 --------------------------------------------------------------------------------
 IMP  #68    Adds version for Apple AArch64 (aka Arm64, i.e. M1 etc.) architecture.
+FIX  #69    Compilation aborts if output filename is too long.
 
 
 2022-10-15  2.11.4

--- a/intervalmusiccompositor.commons/src/main/java/ch/retorte/intervalmusiccompositor/commons/platform/LinuxPlatform.java
+++ b/intervalmusiccompositor.commons/src/main/java/ch/retorte/intervalmusiccompositor/commons/platform/LinuxPlatform.java
@@ -75,4 +75,10 @@ public class LinuxPlatform extends Platform {
     public String getUpgradeExecutableName() {
         return "intervalmusiccompositor-updater";
     }
+
+    @Override
+    public int getMaximumPathLength() {
+        /* Note: This is actually the maximum filename length. */
+        return 255;
+    }
 }

--- a/intervalmusiccompositor.commons/src/main/java/ch/retorte/intervalmusiccompositor/commons/platform/MacOsxPlatform.java
+++ b/intervalmusiccompositor.commons/src/main/java/ch/retorte/intervalmusiccompositor/commons/platform/MacOsxPlatform.java
@@ -40,4 +40,9 @@ public class MacOsxPlatform extends Platform {
     public String getUpgradeExecutableName() {
         return "intervalmusiccompositor-updater";
     }
+
+    @Override
+    public int getMaximumPathLength() {
+        return 255;
+    }
 }

--- a/intervalmusiccompositor.commons/src/main/java/ch/retorte/intervalmusiccompositor/commons/platform/Platform.java
+++ b/intervalmusiccompositor.commons/src/main/java/ch/retorte/intervalmusiccompositor/commons/platform/Platform.java
@@ -304,4 +304,9 @@ public abstract class Platform {
         return new File(getBinaryPathFrom(distributionRoot).getAbsolutePath() + File.separator + getProgramExecutableName());
     }
 
+    /**
+     * Returns the maximum length of an absolute path.
+     */
+    public abstract int getMaximumPathLength();
+
 }

--- a/intervalmusiccompositor.commons/src/main/java/ch/retorte/intervalmusiccompositor/commons/platform/WindowsPlatform.java
+++ b/intervalmusiccompositor.commons/src/main/java/ch/retorte/intervalmusiccompositor/commons/platform/WindowsPlatform.java
@@ -52,4 +52,9 @@ public class WindowsPlatform extends Platform {
     public String getUpgradeExecutableName() {
         return "intervalmusiccompositor-updater.exe";
     }
+
+    @Override
+    public int getMaximumPathLength() {
+        return 260;
+    }
 }

--- a/intervalmusiccompositor.core/src/main/java/ch/retorte/intervalmusiccompositor/compilation/CompilationGenerator.java
+++ b/intervalmusiccompositor.core/src/main/java/ch/retorte/intervalmusiccompositor/compilation/CompilationGenerator.java
@@ -1,5 +1,7 @@
 package ch.retorte.intervalmusiccompositor.compilation;
 
+import ch.retorte.intervalmusiccompositor.commons.platform.Platform;
+import ch.retorte.intervalmusiccompositor.commons.platform.PlatformFactory;
 import ch.retorte.intervalmusiccompositor.core.bundle.CoreBundleProvider;
 import ch.retorte.intervalmusiccompositor.commons.bundle.MessageFormatBundle;
 import ch.retorte.intervalmusiccompositor.model.audiofile.IAudioFile;
@@ -43,6 +45,9 @@ public class CompilationGenerator implements Runnable {
   private static final String FILENAME_PREFIX_PART_DELIMITER = "_";
   private static final String FILENAME_PREFIX_ITERATION_DELIMITER = "-";
 
+
+  private final Platform platform = new PlatformFactory().getPlatform();
+
   private final MessageFormatBundle bundle = new CoreBundleProvider().getBundle();
 
   private final ArrayList<TaskFinishListener> listeners = newArrayList();
@@ -56,9 +61,9 @@ public class CompilationGenerator implements Runnable {
   private final ArrayList<IAudioFile> breakPlaylistCandidates = newArrayList();
   private final ArrayList<IAudioFile> badSoundFiles = newArrayList();
 
-  private String playlist_outfile;
+  private String playlistReportFilename;
   private String correctedOutputPath;
-  private String outfile_prefix;
+  private String musicFilenameWithoutExtension;
 
   private Playlist playlist;
 
@@ -161,14 +166,36 @@ public class CompilationGenerator implements Runnable {
 
   private void compileOutputParameters() {
     correctedOutputPath = compilationParameters.getOutputPath();
-    if (correctedOutputPath == null || correctedOutputPath.equals("")) {
+    if (correctedOutputPath == null || correctedOutputPath.isEmpty()) {
       correctedOutputPath = bundle.getString("imc.workPath");
     }
     createIfNonExisting(new File(correctedOutputPath));
 
-    String identification_prefix = createMusicAndBreakPatternPrefixWith(compilationParameters.getMusicPattern(), compilationParameters.getBreakPattern(), compilationParameters.getIterations());
-    playlist_outfile = identification_prefix + FILENAME_PART_DELIMITER + bundle.getString("imc.outfile.playlist.suffix");
-    outfile_prefix = identification_prefix + FILENAME_PART_DELIMITER + bundle.getString("imc.outfile.sound.infix");
+    final String playlistFileSuffix = bundle.getString("imc.outfile.playlist.suffix");
+    final String musicFileSuffix = bundle.getString("imc.outfile.sound.infix");
+
+    final int remainingPathLength = platform.getMaximumPathLength() - correctedOutputPath.length() - Math.max(playlistFileSuffix.length(), musicFileSuffix.length()) - 3;
+
+    if (remainingPathLength <= 0) {
+      throw new CompilationException(bundle.getString("ui.error.filetoolong"), null);
+    }
+
+    final String originalFilenamePrefix = createMusicAndBreakPatternPrefixWith(compilationParameters.getMusicPattern(), compilationParameters.getBreakPattern(), compilationParameters.getIterations());
+    final String truncatedFilenamePrefix = truncateIfLongerThan(remainingPathLength, originalFilenamePrefix);
+
+    playlistReportFilename = truncatedFilenamePrefix + FILENAME_PART_DELIMITER + playlistFileSuffix;
+    musicFilenameWithoutExtension = truncatedFilenamePrefix + FILENAME_PART_DELIMITER + musicFileSuffix;
+  }
+
+  private String truncateIfLongerThan(int length, String s) {
+    if (s != null && length < s.length() ) {
+      final String truncated = s.substring(0, length - 1) + "x";
+      addDebugMessage(String.format("Original file name\n'%s' (length %d)\ntoo long; truncated to:\n'%s' (length %d).", s, s.length(), truncated, truncated.length()));
+      return truncated;
+    }
+    else {
+      return s;
+    }
   }
 
   private void createIfNonExisting(File f) {
@@ -231,7 +258,7 @@ public class CompilationGenerator implements Runnable {
   }
 
   private void createPlaylistReport() {
-    File reportFile = new File(correctedOutputPath + "/" + playlist_outfile);
+    File reportFile = new File(correctedOutputPath + "/" + playlistReportFilename);
     String playlistReport = new PlaylistReport(applicationData).generateReportFor(playlist, badSoundFiles);
 
     try (OutputStreamWriter outputStreamWriter = new OutputStreamWriter(new FileOutputStream(reportFile), UTF_8)) {
@@ -280,7 +307,7 @@ public class CompilationGenerator implements Runnable {
 
   private void writeOutputFile() {
     try (FileInputStream compilationData = new FileInputStream(compilationDataFile)) {
-      outputGenerator.generateOutput(compilationData, compilationDataSize, correctedOutputPath, outfile_prefix, compilationParameters.getEncoderIdentifier(), new InertProgressListener() {
+      outputGenerator.generateOutput(compilationData, compilationDataSize, correctedOutputPath, musicFilenameWithoutExtension, compilationParameters.getEncoderIdentifier(), new InertProgressListener() {
 
         @Override
         protected void onProgressChange(int percent) {

--- a/intervalmusiccompositor.core/src/main/resources/core_imc_de.properties
+++ b/intervalmusiccompositor.core/src/main/resources/core_imc_de.properties
@@ -11,6 +11,7 @@ ui.error.general.introduction = Folgender Fehler trat auf.\n\r
 ui.error.memory.introduction = Das Programm ben√∂tigt f√ºr einen Vorgang mehr Hauptspeicher (RAM) als es zugewiesen erhielt.\n\r
 ui.error.sound.introduction = Eine Musikdatei verursachte einen Fehler, von dem sich das Programm nicht erholen kann.\n\r
 ui.error.soundeffectformat.introduction = Das Datenformat des Soundeffekts ist unbekannt.\n\r
+ui.error.filetoolong = Das gew‰hlte Ausgabeverzeichnis ist zu lang, es kann keine Datei mehr erstellt werden. W‰hle ein anderes Ausgabeverzeichnis.
 
 # Status of the compilation
 compilation.status.fresh =

--- a/intervalmusiccompositor.core/src/main/resources/core_imc_en.properties
+++ b/intervalmusiccompositor.core/src/main/resources/core_imc_en.properties
@@ -11,6 +11,7 @@ ui.error.general.introduction = There was a problem!\n\r
 ui.error.memory.introduction = The program gets not enough memory (RAM) to process your request!\n\r
 ui.error.sound.introduction = There was a problem with at least one of the sound files!\n\r
 ui.error.soundeffectformat.introduction = The sound effect has an unknown data format.\n\r
+ui.error.filetoolong = The chosen output directory is too long, there is no space left for creating a new file. Please choose another directory.
 
 # Status of the compilation
 compilation.status.fresh =

